apexlog/apexlog.go

```diff
@@ -0,0 +1,143 @@
+// the flag.Value interface. The -stderrthreshold flag is of type severity and
+// should be modified only through the flag.Value interface. The values match
+// the corresponding constants in C++.
+type severity int8
+
+// These constants identify the log levels in order of increasing severity.
+// A message written to a high-severity log file is also written to each
+// lower-severity log file.
+const (
+	debugLog severity = iota
+	infoLog
+	warningLog
+	errorLog
+	fatalLog
+)
+
+var severityName = []string{
+	debugLog:   "debug",
+	infoLog:    "info",
+	warningLog: "warning",
+	errorLog:   "error",
+	fatalLog:   "fatal",
+}
+
+func severityByName(s string) (severity, bool) {
+	for i, name := range severityName {
+		if name == strings.ToLower(s) {
+			return severity(i), true
+		}
+	}
+	return 0, false
+}
+
+var gobrakeSeverity = errorLog
+
+// SetGobrakeSeverity sets minimum log severity that will be sent to Airbrake.
+//
+// Valid names are "debug" "info", "warning", "error", and "fatal".  If the name is not
+// recognized, "error" severity is used.
+func setGobrakeSeverity(severity string) {
+	name, ok := severityByName(severity)
+	if ok {
+		gobrakeSeverity = name
+	}
+}
+
+// Gobrake is an instance of Airbrake Go Notifier that is used to send
+// logs to Airbrake.
+var Gobrake *gobrake.Notifier
+
+// Handler implementation.
+type Handler struct {
+	Gobrake *gobrake.Notifier
+}
+
+// New Apex Logs handler for airbrake.
+//
+// Valid severity levels are "debug" "info", "warning", "error", and "fatal".  If the name is not
+// recognized, "error" severity is used.
+func New(notifier *gobrake.Notifier, severity string) *Handler {
+	setGobrakeSeverity(severity)
+	var v Handler
+	Gobrake = notifier
+	v.Gobrake = notifier
+	return &v
+}
+
+// HandleLog method is used for sendi notices to airbrake.
+func (h *Handler) HandleLog(e *log.Entry) error {
+	notifyAirbrake(2, severity(e.Level), e.Message)
+
+	return nil
+}
+
+type requester interface {
+	Request() *http.Request
+}
+
+// Implemented by context.Context
+type valuer interface {
+	Value(key interface{}) interface{}
+}
+
+func notifyAirbrake(depth int, s severity, args ...interface{}) {
+	if Gobrake == nil {
+		return
+	}
+	if s < gobrakeSeverity {
+		return
+	}
+
+	msg := fmt.Sprint(args...)
+
+	var theErr error
+	var req *http.Request
+	var values valuer
+	for _, arg := range args {
+		if v, ok := arg.(error); ok {
+			theErr = v
+		} else if v, ok := arg.(*http.Request); ok {
+			req = v
+		} else if v, ok := arg.(requester); ok {
+			req = v.Request()
+		} else if v, ok := arg.(valuer); ok {
+			values = v
+		}
+	}
+
+	if values != nil && req == nil {
+		req, _ = values.Value("request").(*http.Request)
+	}
+
+	var notice *gobrake.Notice
+	if theErr != nil {
+		notice = Gobrake.Notice(theErr, req, depth)
+		notice.Errors[0].Message = msg
+	} else {
+		notice = Gobrake.Notice(msg, req, depth)
+	}
+	notice.Context["severity"] = severityName[s]
```

Comment:
If `s` implemented `String()`, we could call:
```suggestion
	notice.Context["severity"] = s.String()
```